File: python_port/tools/export_nodo_summary.m
function export_nodo_summary(matFile, outCsv)
%EXPORT_NODO_SUMMARY Summarize a MATLAB Nodo_*.mat file to CSV for cross-checking
%   against the Python port's output (see python_port/tools/compare_summaries.py).
%
%   export_nodo_summary('data/interim/Dati10/Nodo_Dati10_20260115_20260116.mat', ...
%                        'python_port/tools/_matlab_summary.csv')
%
%   Writes one row per sensor: ID, Label, NumSamples, FirstTime, LastTime,
%   MeanPressure, StdPressure, MinPressure, MaxPressure.

if nargin < 2
    [d, b, ~] = fileparts(matFile);
    outCsv = fullfile(d, [b '_summary.csv']);
end

S = load(matFile);
if isfield(S, 'Nodo')
    Nodo = S.Nodo;
elseif isfield(S, 'Nodo_filtered')
    Nodo = S.Nodo_filtered;
else
    error('export_nodo_summary:MissingVar', 'File has neither Nodo nor Nodo_filtered.');
end

n = numel(Nodo);
ID = strings(n,1); Label = strings(n,1);
NumSamples = zeros(n,1);
FirstTime = strings(n,1); LastTime = strings(n,1);
MeanPressure = nan(n,1); StdPressure = nan(n,1);
MinPressure = nan(n,1); MaxPressure = nan(n,1);

for k = 1:n
    ID(k) = string(Nodo(k).ID);
    Label(k) = string(Nodo(k).Label);
    p = Nodo(k).Pressure;
    t = Nodo(k).Time;
    NumSamples(k) = numel(p);
    if ~isempty(t)
        FirstTime(k) = string(datestr(t(1), 'yyyy-mm-ddTHH:MM:SS.FFF'));
        LastTime(k)  = string(datestr(t(end), 'yyyy-mm-ddTHH:MM:SS.FFF'));
    end
    if ~isempty(p)
        MeanPressure(k) = mean(p);
        StdPressure(k)  = std(p);
        MinPressure(k)  = min(p);
        MaxPressure(k)  = max(p);
    end
end

T = table(ID, Label, NumSamples, FirstTime, LastTime, MeanPressure, StdPressure, MinPressure, MaxPressure);
writetable(T, outCsv);
fprintf('Wrote summary for %d sensors to %s\n', n, outCsv);
end

File: python_port/tools/export_nodo_summary.py
"""Summarize a Python-port Nodo_*.pkl file to CSV, in the same format as
export_nodo_summary.m, so the two can be diffed by compare_summaries.py.

Usage:
    python -m python_port.tools.export_nodo_summary <nodo.pkl> [out.csv]
"""
from __future__ import annotations

import pickle
import sys
from pathlib import Path

import numpy as np
import pandas as pd


def export_nodo_summary(pkl_path: Path, out_csv: Path = None) -> Path:
    pkl_path = Path(pkl_path)
    if out_csv is None:
        out_csv = pkl_path.with_name(pkl_path.stem + "_summary.csv")

    with open(pkl_path, "rb") as f:
        nodo = pickle.load(f)

    rows = []
    for s in nodo:
        p = np.asarray(s["Pressure"], dtype=float)
        t = np.asarray(s["Time"])
        rows.append({
            "ID": str(s["ID"]),
            "Label": str(s["Label"]),
            "NumSamples": len(p),
            "FirstTime": str(t[0]) if len(t) else "",
            "LastTime": str(t[-1]) if len(t) else "",
            "MeanPressure": float(np.mean(p)) if len(p) else np.nan,
            # ddof=1 to match MATLAB's std() default (normalizes by N-1)
            "StdPressure": float(np.std(p, ddof=1)) if len(p) > 1 else np.nan,
            "MinPressure": float(np.min(p)) if len(p) else np.nan,
            "MaxPressure": float(np.max(p)) if len(p) else np.nan,
        })

    df = pd.DataFrame(rows)
    df.to_csv(out_csv, index=False)
    print(f"Wrote summary for {len(df)} sensors to {out_csv}")
    return out_csv


def main(argv=None):
    argv = argv if argv is not None else sys.argv[1:]
    if not argv:
        print(__doc__)
        sys.exit(1)
    pkl_path = Path(argv[0])
    out_csv = Path(argv[1]) if len(argv) > 1 else None
    export_nodo_summary(pkl_path, out_csv)


if __name__ == "__main__":
    main()
